modules/runtime/nuxeo-runtime-migration/src/main/java/org/nuxeo/runtime/migration/MigrationServiceImpl.java

```diff
@@ -529,4 +533,55 @@ protected void atomic(String id, Consumer<KeyValueStore> consumer) {
         throw new RuntimeException("Cannot lock for write migration: " + id + ", already locked: " + currentLock);
     }
 
+    @Override
+    public Migration getMigration(String id) {
+        MigrationDescriptor descriptor = getDescriptor(XP_CONFIG, id);
+        return new Migration(id, descriptor.getDescription(), descriptor.getDescriptionLabel(), getStatus(id),
+                descriptor.getSteps());
+    }
+
+    @Override
+    public List<Migration> getMigrations() {
+        var descriptors = getDescriptors(XP_CONFIG);
+        return descriptors.stream().map(d -> getMigration(d.getId())).collect(Collectors.toList());
+    }
+
+    @Override
+    public void run(String migrationId) {
+        if(getStatus(migrationId) == null) {
+            probeAndSetState(migrationId);
+        }
+        MigrationDescriptor descriptor = getDescriptor(XP_CONFIG, migrationId);
+        Set<MigrationStepDescriptor> availableSteps = getAvailableSteps(descriptor);
+        if(availableSteps.size() != 1) {
+            throw new IllegalStateException(String.format("Migration %s needs to have exactly one step to run", migrationId));
+        } else {
+            runStep(migrationId, availableSteps.iterator().next().getId());
+        }
+    }
+
+    protected Set<MigrationStepDescriptor> getAvailableSteps(MigrationDescriptor descriptor) {
+        MigrationStatus status = getStatus(descriptor.getId());
+        return descriptor.getSteps()
+                         .entrySet()
+                         .stream()
+                         .filter(e -> e.getValue().getFromState().equals(status.getState()))
+                         .map(Entry::getValue)
+                         .collect(Collectors.toSet());
+    }
+
+    @Override
+    public boolean await(Duration duration) throws InterruptedException {
+        long deadline = System.nanoTime() + duration.toNanos();
+        for (;;) {
+            if (executor.getActiveCount() == 0) {
+                return true;
+            }
+            Thread.sleep(100);
+            if (deadline < System.nanoTime()) {
+                return false;
+            }
+        }
+    }
```

Comment:
See my comment on the interface.

For the cluster aspect, if you don't reach the node doing the migration, you won't await I think.

Reply:
followed https://github.com/nuxeo/nuxeo/pull/4027#discussion_r428223297
The only public method left is `MigrationFeature#initialize` but shouldn't be a problem since it's only used for test

Reply:
After inputing it, I have the following problem https://github.com/nuxeo/nuxeo/pull/4027#discussion_r430345263